### PR TITLE
Don't skip fixtures that are substrings of params

### LINKED ISSUE #736: Parametrized argument named similarly to fixture (contains it) breaks that fixture's parametrization
Originally reported by: **Jacobo Giralt (BitBucket: [jgiralt](http://bitbucket.org/jgiralt), GitHub: [jgiralt](http://github.com/jgiralt))**

----------------------------------------

```
#!python

import pytest

@pytest.fixture(params=[1,2,3])
def foo(request):
    return request.param

@pytest.mark.parametrize('foobar', [4,5,6])
def test_issue(foo, foobar):
    assert True
```

```
#!python

request = <SubRequest 'foo' for <Function 'test_issue[4]'>>

    @pytest.fixture(params=[1,2,3])
    def foo(request):
>       return request.param
AttributeError: SubRequest instance has no attribute 'param'
```

The issue seems reproducible as long as the parametrized test argument contains the fixture name as a substring, e.g. foobar, barfoo, foox, xfoo, disappears as soon as I modify the fixture name to anything else.

This same code works in pytest-2.6.4 and fails for 2.7.

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/736

### PR DESCRIPTION
**Pull request by untitaker**
Fix #736

### CODE REVIEW

File: _pytest/python.py [hunk 1]
```diff
@@ -1777,7 +1777,7 @@ def pytest_generate_tests(self, metafunc):
                 if fixturedef.params is not None:
                     func_params = getattr(getattr(metafunc.function, 'parametrize', None), 'args', [[None]])
                     # skip directly parametrized arguments
-                    if argname not in func_params and argname not in func_params[0]:
```
Comment (untitaker) [inline _pytest/python.py]: I didn't figure out why this line is needed. It has been introduced with https://bitbucket.org/pytest-dev/pytest/pull-requests/257/allow-to-override-parametrized-fixtures

I suppose it's due to strings like ´value,othervalue`, but it doesn't seem like it's actually needed.

**Comment by untitaker** [comment]
cc @bubenkoff

File: testing/python/collect.py [hunk 2]
```diff
@@ -412,9 +412,18 @@ def value():
...
                 assert value == 'overridden'
+
+            @pytest.mark.parametrize('somevalue', ['overridden'])
+            def test_not_overridden(value, somevalue):
+                assert value == 'value'
+                assert somevalue == 'overridden'
+
+            @pytest.mark.parametrize('other,value', [('foo', 'overridden')])
+            def test_overridden_via_multiparam(other, value):
+                assert value == 'overridden'
```
Comment (nicoddemus) [inline testing/python/collect.py]: Just for completeness, can you add a `assert other == 'foo'` here?

**Comment by nicoddemus** [comment]
Thanks for the PR @untitaker! :smile:

Could you please also add a note to the `CHANGELOG` (make sure to add a sentence like `"Thanks to Markus Unterwaditzer for the PR"`) and add yourself to `AUTHORS`? Thanks!

**Comment by untitaker** [comment]
Done. I'm not really sure what I did was the right fix though -- I think this should be confirmed somehow.

**Comment by nicoddemus** [comment]
Thanks! I'm not familiar with that part of the code, but based from the tests it seems the change is OK to me.

Let's wait for @bubenkoff's feedback. :smile:

File: CHANGELOG [hunk 3]
```diff
@@ -1,10 +1,14 @@
 2.8.0.dev (compared to 2.7.X)
 -----------------------------
 
+- Fix issue736: Fix a bug where fixture params would be discarded when combined
+  with parametrization markers.
+  Thanks to Markus Unterwaditzer for the PR.
```
Comment (nicoddemus) [inline CHANGELOG]: I think you forgot to add yourself to `AUTHORS` too.

**Comment by untitaker** [comment]
Added myself and also resorted list.

**Comment by nicoddemus** [comment]
Thanks! :beers

**Comment by RonnyPfannschmidt** [comment]
The code should use a string split there I think, and the implementation seems incomplete to begin with

this one seems tricky

**Comment by untitaker** [comment]
The thing is, I couldn't craft a testcase that fails with this patch.

On 8 August 2015 06:22:41 CEST, Ronny Pfannschmidt <notifications@github.com> wrote:
>The code should use a string split there I think, and the
>implementation seems incomplete to begin with
>
>this one seems tricky
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/pytest-dev/pytest/pull/926#issuecomment-128907872

-- 
Sent from my phone. Please excuse my brevity.

**Comment by RonnyPfannschmidt** [comment]
since its an improvements i'm + 1 on merging it anyway, but i suspect we will find an issue at some point in future

**Comment by untitaker** [comment]
If you like, I could add an extra assertion:

    assert not any(',' in x for x in func_params)

**Comment by RonnyPfannschmidt** [comment]
i don't think that would help, i'd really love to have marker signatures at that point

func params is the argument tuple of the marker and it should check the first argument and always turn it into a tuple of names for that particular check (i.e. if its a string, do a ','. split, then do the in check

**Comment by untitaker** [comment]
My point is that this particular piece of code apparently never gets to see a list of strings that contain commas. I can add that assertion and the tests still pass.

**Comment by untitaker** [comment]
Rebased to resolve conflicts in CHANGELOG.

**Comment by untitaker** [comment]
Anything else needed to get this merged?

On 8 August 2015 13:41:26 CEST, Ronny Pfannschmidt <notifications@github.com> wrote:
>i don't think that would help, i'd really love to have marker
>signatures at that point
>
>func params is the argument tuple of the marker and it should check the
>first argument and always turn it into a tuple of names for that
>particular check (i.e. if its a string, do a ','. split, then do the in
>check
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/pytest-dev/pytest/pull/926#issuecomment-128970169

-- 
Sent from my phone. Please excuse my brevity.

**Comment by nicoddemus** [comment]
Looks :+1: to me!

@RonnyPfannschmidt, if you're OK with it, could you please merge this? :smile:

**Comment by RonnyPfannschmidt** [comment]
ok on the merge, will pull in once im at home

github really needs a merge strategy for CHANGELOG

**Comment by The-Compiler** [comment]
If GitHub reads `.gitattributes`, we could add something like `CHANGELOG merge=union` to it. I'll experiment with it later today if I have time.

**Comment by bubenkoff** [comment]
nice!

On 10/08/2015, Florian Bruhin <notifications@github.com> wrote:
> If GitHub reads `.gitattributes`, we could add something like `CHANGELOG
> merge=union` to it. I'll experiment with it later today if I have time.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pytest-dev/pytest/pull/926#issuecomment-129349434


-- 
Anatoly Bubenkov

**Comment by RonnyPfannschmidt** [comment]
@The-Compiler great idea :+1:

**Comment by nicoddemus** [comment]
> @The-Compiler great idea :+1:

:+1:

**Comment by nicoddemus** [comment]
I did a quick test in `pytest-qt` by adding a `.gitattributes` like @The-Compiler said, and it seems to work fine in my local tests. :smile:

**Comment by nicoddemus** [comment]
Hmm I just pushed a `.gitattributes`to `master` in 681e502, but GitHub is still marking this PR as conflicted...

**Comment by untitaker** [comment]
Rebase worked fine though.

**Comment by nicoddemus** [comment]
Oh well, let's keep an eye on future PR at least. :smile:

**Comment by untitaker** [comment]
Just temporarily undid the rebase and it indeed seems that GitHub doesn't care about the .gitattributes file. :(

**Comment by nicoddemus** [comment]
:expressionless:

**Comment by nicoddemus** [comment]
Thanks again @untitaker!

**Pull request merged**